presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +582,193 @@ private static NestedField nestedField(String name)
+                        for (int callArgumentIndex = 0; callArgumentIndex < call.getArguments().size(); callArgumentIndex++) {
+                            // Since context is global during the traversal of all the nodes in expression tree, we need to pass lambda subfields only to those  function
+                            // arguments that they relate to.
+                            if (argumentIndexToLambdaSubfieldsMap.containsKey(callArgumentIndex)) {
+                                context.setLambdaSubfields(Optional.of(argumentIndexToLambdaSubfieldsMap.get(callArgumentIndex)));
+                            }
+                            else {
+                                context.setLambdaSubfields(Optional.empty());
+                            }
+                            call.getArguments().get(callArgumentIndex).accept(this, context);
+                        }
+
+                        // When we are done with inner calls (child nodes) we need to restore lambda subfields we received from parent expression to handle such situations like
+                        // in example below
+                        // SELECT * FROM my_table WHERE ANY_MATCH(column1, x -> x.ds > '2023-01-01') AND ALL_MATCH(column2, x -> STRPOS(x.comment,  'Presto') > 0)
+                        // After we are done with ANY_MATCH, we need to restore the lambda subfields to what we received from parent node 'AND' so that it does not collide with
+                        // lambda subfields of ALL_MATCH function.
+                        context.setLambdaSubfields(lambdaSubfieldsOriginal);
+                        return null;
+                    }
                 }
-                else {
+                call.getArguments().forEach(argument -> argument.accept(this, context));
+                return null;
+            }
+
+            private static void addNoneSubfieldIfNoAccessedSubfieldsFound(CallExpression call, Map<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMap)
+            {
+                for (Integer callArgumentIndex : argumentIndexToLambdaSubfieldsMap.keySet()) {
+                    if (!argumentIndexToLambdaSubfieldsMap.get(callArgumentIndex).isEmpty()) {
+                        continue;
+                    }
+                    RowExpression argument = call.getArguments().get(callArgumentIndex);
+                    if ((argument.getType() instanceof ArrayType && ((ArrayType) argument.getType()).getElementType() instanceof RowType) ||
+                            (argument.getType() instanceof MapType && ((MapType) argument.getType()).getValueType() instanceof RowType)) {
+                        argumentIndexToLambdaSubfieldsMap.put(callArgumentIndex, ImmutableSet.of(new Subfield("", ImmutableList.of(allSubscripts(), noneSubfield()))));
+                    }
+                }
+            }
+
+            private boolean addLambdaSubfields(CallExpression call, Context context, LambdaDescriptor lambdaDescriptor,
```

Comment:
we avoid modifying method arguments (e.g. adding elements to list or maps).  Instead of returning a boolean to indicate whether any arguments have been added, have this function return any subfields that need to be added and have the caller add them where needed.

Reply:
Fixed